--- a/src/Tools/VSCode/src/lsp.scala
+++ b/src/Tools/VSCode/src/lsp.scala
@@ -667,6 +667,40 @@
       }
   }
 
+  object Theory_Status extends Request0("PIDE/theory_status") {
+    def reply(id: Id, theories: List[JSON.T]): JSON.T =
+      ResponseMessage(id, Some(JSON.Object("theories" -> theories)))
+  }
+
+  object Cancel_Execution extends Request0("PIDE/cancel_execution") {
+    def reply(id: Id): JSON.T =
+      ResponseMessage(id, Some(JSON.Object("cancelled" -> true)))
+  }
+
+  object Command_At_Position extends RequestTextDocumentPosition("PIDE/command_at_position") {
+    def reply(id: Id, result: Option[(String, Line.Range)]): JSON.T = {
+      val res =
+        result match {
+          case Some((source, range)) =>
+            JSON.Object("source" -> source, "range" -> Range(range))
+          case None => JSON.Object()
+        }
+      ResponseMessage(id, Some(res))
+    }
+  }
+
+  object Output_At_Position extends RequestTextDocumentPosition("PIDE/output_at_position") {
+    def reply(id: Id, result: Option[(String, Line.Range, String)]): JSON.T = {
+      val res =
+        result match {
+          case Some((source, range, content)) =>
+            JSON.Object("source" -> source, "range" -> Range(range), "content" -> content)
+          case None => JSON.Object()
+        }
+      ResponseMessage(id, Some(res))
+    }
+  }
+
 
   /* preview */
 
