--- a/src/Tools/VSCode/src/lsp.scala
+++ b/src/Tools/VSCode/src/lsp.scala
@@ -667,6 +667,80 @@
       }
   }
 
+  object Theory_Status extends Request0("PIDE/theory_status") {
+    def reply(id: Id, theories: List[JSON.T]): JSON.T =
+      ResponseMessage(id, Some(JSON.Object("theories" -> theories)))
+  }
+
+  object Cancel_Execution extends Request0("PIDE/cancel_execution") {
+    def reply(id: Id): JSON.T =
+      ResponseMessage(id, Some(JSON.Object("cancelled" -> true)))
+  }
+
+  object Command_At_Position extends RequestTextDocumentPosition("PIDE/command_at_position") {
+    def reply(id: Id, result: Option[(String, Line.Range)]): JSON.T = {
+      val res =
+        result match {
+          case Some((source, range)) =>
+            JSON.Object("source" -> source, "range" -> Range(range))
+          case None => JSON.Object()
+        }
+      ResponseMessage(id, Some(res))
+    }
+  }
+
+  object Output_At_Position extends RequestTextDocumentPosition("PIDE/output_at_position") {
+    def reply(id: Id, result: Option[(String, Line.Range, String)]): JSON.T = {
+      val res =
+        result match {
+          case Some((source, range, content)) =>
+            JSON.Object("source" -> source, "range" -> Range(range), "content" -> content)
+          case None => JSON.Object()
+        }
+      ResponseMessage(id, Some(res))
+    }
+  }
+
+  object Symbols extends Request0("PIDE/symbols") {
+    def reply(id: Id, content: String): JSON.T =
+      ResponseMessage(id, Some(JSON.Object("content" -> content)))
+  }
+
+  object Find_Theorems_Request {
+    def unapply(json: JSON.T): Option[(String, List[String])] =
+      json match {
+        case Notification("PIDE/find_theorems_request", Some(params)) =>
+          for {
+            token <- JSON.string(params, "token")
+            query <- JSON.string(params, "query")
+            limit <- JSON.string(params, "limit")
+            allow_dups <- JSON.string(params, "allow_dups")
+          } yield (token, List(limit, allow_dups, query))
+        case _ => None
+      }
+  }
+
+  object Find_Theorems_Status {
+    def apply(token: String, status: String): JSON.T =
+      Notification("PIDE/find_theorems_status",
+        JSON.Object("token" -> token, "status" -> status))
+  }
+
+  object Find_Theorems_Output {
+    def apply(token: String, content: String): JSON.T =
+      Notification("PIDE/find_theorems_output",
+        JSON.Object("token" -> token, "content" -> content))
+  }
+
+  object Find_Theorems_Cancel {
+    def unapply(json: JSON.T): Option[String] =
+      json match {
+        case Notification("PIDE/find_theorems_cancel", Some(params)) =>
+          JSON.string(params, "token")
+        case _ => None
+      }
+  }
+
 
   /* preview */
 
