--- a/src/Pure/Isar/attrib.ML
+++ b/src/Pure/Isar/attrib.ML
@@ -605,6 +605,7 @@
   register_config_bool Printer.show_brackets #>
   register_config_bool Printer.show_sorts #>
   register_config_bool Printer.show_types #>
+  register_config_bool Printer.show_types_nv #>
   register_config_bool Printer.show_markup #>
   register_config_bool Printer.show_structs #>
   register_config_bool Printer.show_question_marks #>
