--- a/etc/options
+++ b/etc/options
@@ -55,6 +55,8 @@
 
 option show_types : bool = false for content
   -- "show type constraints when printing terms"
+option show_types_nv : bool = false for content
+  -- "effective only when `show_types` is on. Disables the type annotation of free/fixed variables."
 option show_sorts : bool = false for content
   -- "show sort constraints when printing types"
 option show_brackets : bool = false for content
