--- a/src/Tools/VSCode/src/lsp.scala	2026-05-26 15:23:27.416155517 +0800
+++ b/src/Tools/VSCode/src/lsp.scala	2026-05-26 15:24:45.189747770 +0800
@@ -672,6 +672,11 @@
       ResponseMessage(id, Some(JSON.Object("theories" -> theories)))
   }
 
+  object Cancel_Execution extends Request0("PIDE/cancel_execution") {
+    def reply(id: Id): JSON.T =
+      ResponseMessage(id, Some(JSON.Object("cancelled" -> true)))
+  }
+
 
   /* preview */
 
