--- a/src/Tools/VSCode/src/lsp.scala	2026-05-26 15:23:33.792225001 +0800
+++ b/src/Tools/VSCode/src/lsp.scala	2026-05-26 15:24:52.993851031 +0800
@@ -553,6 +553,11 @@
       ResponseMessage(id, Some(JSON.Object("theories" -> theories)))
   }
 
+  object Cancel_Execution extends Request0("PIDE/cancel_execution") {
+    def reply(id: Id): JSON.T =
+      ResponseMessage(id, Some(JSON.Object("cancelled" -> true)))
+  }
+
 
   /* preview */
 
