--- a/src/Pure/PIDE/protocol.ML	2026-05-26 15:23:33.789688405 +0800
+++ b/src/Pure/PIDE/protocol.ML	2026-05-26 15:24:25.551488589 +0800
@@ -93,6 +93,10 @@
     (fn [] => Execution.discontinue ());
 
 val _ =
+  Protocol_Command.define "Document.cancel_execution"
+    (fn [] => Execution.cancel_execution ());
+
+val _ =
   Protocol_Command.define "Document.cancel_exec"
     (fn [exec_id] => Execution.cancel (Document_ID.parse exec_id));
 
