--- a/src/Tools/VSCode/src/lsp.scala	2026-05-25 22:01:41.185857773 +0800
+++ b/src/Tools/VSCode/src/lsp.scala	2026-05-25 22:01:57.388497438 +0800
@@ -667,6 +667,11 @@
       }
   }
 
+  object Theory_Status extends Request0("PIDE/theory_status") {
+    def reply(id: Id, theories: List[JSON.T]): JSON.T =
+      ResponseMessage(id, Some(JSON.Object("theories" -> theories)))
+  }
+
 
   /* preview */
 
