--- a/src/Pure/PIDE/execution.ML	2026-05-26 15:23:33.786810203 +0800
+++ b/src/Pure/PIDE/execution.ML	2026-05-26 15:24:10.658292682 +0800
@@ -9,6 +9,7 @@
 sig
   val start: unit -> Document_ID.execution
   val discontinue: unit -> unit
+  val cancel_execution: unit -> unit
   val is_running: Document_ID.execution -> bool
   val active_tasks: string -> Future.task list
   val worker_task_active: bool -> string -> unit
@@ -77,6 +78,13 @@
 fun discontinue () =
   change_state (fn (_, nodes, execs) => (Document_ID.none, nodes, execs));
 
+fun cancel_execution () =
+  let
+    val groups = change_state_result (fn (_, nodes, execs) =>
+      let val groups = Inttab.fold (fn (_, (gs, _)) => fn acc => gs @ acc) execs []
+      in (groups, (Document_ID.none, nodes, execs)) end);
+  in List.app Future.cancel_group groups end;
+
 fun is_running execution_id =
   execution_id = #execution_id (get_state ());
 
