--- a/src/Pure/Syntax/syntax_phases.ML
+++ b/src/Pure/Syntax/syntax_phases.ML
@@ -659,6 +659,7 @@
 fun term_to_ast ctxt trf =
   let
     val show_types = Config.get ctxt show_types orelse Config.get ctxt show_sorts;
+    val show_types_nv = Config.get ctxt show_types_nv;
     val show_markup = Config.get ctxt show_markup;
     val show_consts_markup = Config.get ctxt show_consts_markup;
 
@@ -676,17 +677,17 @@
       (case strip_comb tm of
         (t as Abs _, ts) => Ast.mk_appl (main (Syntax_Trans.abs_tr' ctxt t)) (map main ts)
       | ((c as Const ("_free", _)), Free (x, T) :: ts) =>
-          Ast.mk_appl (variable (c $ Syntax.free x) T) (map main ts)
+          Ast.mk_appl (variable true (c $ Syntax.free x) T) (map main ts)
       | ((c as Const ("_var", _)), Var (xi, T) :: ts) =>
-          Ast.mk_appl (variable (c $ Syntax.var xi) T) (map main ts)
+          Ast.mk_appl (variable false (c $ Syntax.var xi) T) (map main ts)
       | ((c as Const ("_bound", B)), Free (x, T) :: ts) =>
           let
             val X =
               if show_markup andalso not show_types orelse B <> dummyT then T
               else dummyT;
-          in Ast.mk_appl (variable (c $ Syntax.free x) X) (map main ts) end
+          in Ast.mk_appl (variable false (c $ Syntax.free x) X) (map main ts) end
       | (Const ("_idtdummy", T), ts) =>
-          Ast.mk_appl (variable (Syntax.const "_idtdummy") T) (map main ts)
+          Ast.mk_appl (variable false (Syntax.const "_idtdummy") T) (map main ts)
       | (Const (c, T), ts) => constant c T ts
       | (t, ts) => Ast.mk_appl (simple_ast_of ctxt t) (map main ts))
 
@@ -697,9 +698,9 @@
           let val c = Ast.Constant a |> show_const_types ? constrain {clean = true} T
           in Ast.mk_appl c (map main args) end)
 
-    and variable v T =
+    and variable is_v v T =
       simple_ast_of ctxt v
-      |> show_var_types ? constrain {clean = clean_var_types} T;
+      |> (show_var_types andalso (not is_v orelse not show_types_nv)) ? constrain {clean = clean_var_types} T;
   in
     mark_aprop
     #> show_types ? prune_types
